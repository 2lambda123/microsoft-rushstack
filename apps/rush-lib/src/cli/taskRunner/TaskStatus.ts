// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Enumeration defining potential states of a task: not started, executing, or completed
 */
enum TaskStatus {
  Ready = 1,
  Executing = 2,
  Success = 3,
  SuccessWithWarning = 4,
  Skipped = 5,
  Failure = 6,
  Blocked = 7
}
export default TaskStatus;
