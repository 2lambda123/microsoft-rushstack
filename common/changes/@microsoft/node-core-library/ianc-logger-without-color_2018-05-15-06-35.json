{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Add new logging API that can remove console colors.",
      "type": "minor"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "iclanton@users.noreply.github.com"
}