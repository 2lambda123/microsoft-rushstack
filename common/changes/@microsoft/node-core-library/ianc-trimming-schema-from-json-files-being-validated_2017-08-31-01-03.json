{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Trim the $schema property from a JSON object being validated.",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "iclanton@users.noreply.github.com"
}