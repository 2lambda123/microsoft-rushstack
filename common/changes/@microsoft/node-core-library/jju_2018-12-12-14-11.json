{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Use @types/jju not custom typings",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "ecraig12345@users.noreply.github.com"
}