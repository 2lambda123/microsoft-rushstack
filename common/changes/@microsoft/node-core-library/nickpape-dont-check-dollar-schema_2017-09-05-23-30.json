{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Updated JsonSchema to not check the $schema field",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "nickpape@users.noreply.github.com"
}