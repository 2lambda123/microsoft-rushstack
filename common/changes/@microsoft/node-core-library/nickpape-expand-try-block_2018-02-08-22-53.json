{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Fix an issue for LockFiles where not all filesystem operations were wrapped in a try/catch block.",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "nickpape-msft@users.noreply.github.com"
}