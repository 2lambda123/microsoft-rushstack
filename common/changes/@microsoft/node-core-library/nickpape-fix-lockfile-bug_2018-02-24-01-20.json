{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Fix an issue where the LockFile was unable to acquire the lock if the resource dir doesn't exist.",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "nickpape-msft@users.noreply.github.com"
}