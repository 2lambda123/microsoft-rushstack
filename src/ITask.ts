import { IBuildConfig } from './IBuildConfig';

export interface ITask<CONFIG_TYPE> {
  config: (taskConfig: CONFIG_TYPE) => void;

  execute: (config: IBuildConfig) => Promise<any>;

  name?: string;
  buildConfig?: IBuildConfig;
  taskConfig?: CONFIG_TYPE;
}

export default ITask;
