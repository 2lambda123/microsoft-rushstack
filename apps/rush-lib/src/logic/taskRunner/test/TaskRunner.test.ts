import { EOL } from 'os';
import { TaskRunner } from '../TaskRunner';
import { ITaskWriter } from '@microsoft/stream-collator';
import { TaskStatus } from '../TaskStatus';
import { ITaskDefinition } from '../ITask';
import { StringBufferTerminalProvider, Terminal } from '@microsoft/node-core-library';

function createDummyTask(name: string, action?: () => void): ITaskDefinition {
  return {
    name,
    isIncrementalBuildAllowed: false,
    execute: (writer: ITaskWriter) => {
      if (action) {
        action();
      }
      return Promise.resolve(TaskStatus.Success);
    }
  };
}

function checkConsoleOutput(terminalProvider: StringBufferTerminalProvider): void {
  expect(terminalProvider.getOutput()).toMatchSnapshot();
  expect(terminalProvider.getVerbose()).toMatchSnapshot();
  expect(terminalProvider.getWarningOutput()).toMatchSnapshot();
  expect(terminalProvider.getErrorOutput()).toMatchSnapshot();
}

describe('TaskRunner', () => {
  let terminalProvider: StringBufferTerminalProvider;
  let terminal: Terminal;
  let taskRunner: TaskRunner;

  beforeEach(() => {
    terminalProvider = new StringBufferTerminalProvider(true);
    terminal = new Terminal(terminalProvider);
  });

  describe('Constructor', () => {
    it('throwsErrorOnInvalidParallelism', () => {
      expect(() => new TaskRunner(false, 'tequila', false, terminal)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('Dependencies', () => {
    beforeEach(() => {
      taskRunner = new TaskRunner(false, '1', false, terminal);
    });

    it('throwsErrorOnNonExistentTask', () => {
      expect(() => taskRunner.addDependencies('foo', []))
        .toThrowErrorMatchingSnapshot();
    });

    it('throwsErrorOnNonExistentDependency', () => {
      taskRunner.addTask(createDummyTask('foo'));
      expect(() => taskRunner.addDependencies('foo', ['bar']))
        .toThrowErrorMatchingSnapshot();
    });

    it('detectsDependencyCycle', () => {
      taskRunner.addTask(createDummyTask('foo'));
      taskRunner.addTask(createDummyTask('bar'));
      taskRunner.addDependencies('foo', ['bar']);
      taskRunner.addDependencies('bar', ['foo']);
      expect(() => taskRunner.execute()).toThrowErrorMatchingSnapshot();
    });

    it('respectsDependencyOrder', () => {
      const result: Array<string> = [];
      taskRunner.addTask(createDummyTask('two', () => result.push('2')));
      taskRunner.addTask(createDummyTask('one', () => result.push('1')));
      taskRunner.addDependencies('two', ['one']);
      return taskRunner
        .execute()
        .then(() => {
          expect(result.join(',')).toEqual('1,2');
          checkConsoleOutput(terminalProvider);
        })
        .catch(error => fail(error));
    });
  });

  describe('Error logging', () => {
    beforeEach(() => {
      taskRunner = new TaskRunner(false, '1', false, terminal);
    });

    const EXPECTED_FAIL: string = 'Promise returned by execute() resolved but was expected to fail';

    it('printedStderrAfterError', () => {
      taskRunner.addTask({
        name: 'stdout+stderr',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.write('Build step 1' + EOL);
          writer.writeError('Error: step 1 failed' + EOL);
          return Promise.resolve(TaskStatus.Failure);
        }
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          const allMessages: string = terminalProvider.getOutput();
          expect(allMessages).not.toContain('Build step 1');
          expect(allMessages).toContain('Error: step 1 failed');
          checkConsoleOutput(terminalProvider);
        });
    });

    it('printedStdoutAfterErrorWithEmptyStderr', () => {
      taskRunner.addTask({
        name: 'stdout only',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.write('Build step 1' + EOL);
          writer.write('Error: step 1 failed' + EOL);
          return Promise.resolve(TaskStatus.Failure);
        }
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          expect(terminalProvider.getOutput()).toMatch(/Build step 1.*Error: step 1 failed/);
          checkConsoleOutput(terminalProvider);
        });
    });

    it('printedAbridgedStdoutAfterErrorWithEmptyStderr', () => {
      taskRunner.addTask({
        name: 'large stdout only',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.write(`Building units...${EOL}`);
          for (let i: number = 1; i <= 50; i++) {
            writer.write(` - unit #${i};${EOL}`);
          }
          return Promise.resolve(TaskStatus.Failure);
        }
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          expect(terminalProvider.getOutput())
            .toMatch(/Building units.* - unit #1;.* - unit #3;.*lines omitted.* - unit #48;.* - unit #50;/);
          checkConsoleOutput(terminalProvider);
        });
    });

    it('preservedLeadingBlanksButTrimmedTrailingBlanks', () => {
      taskRunner.addTask({
        name: 'large stderr with leading and trailing blanks',
        isIncrementalBuildAllowed: false,
        execute: (writer: ITaskWriter) => {
          writer.writeError(`List of errors:  ${EOL}`);
          for (let i: number = 1; i <= 50; i++) {
            writer.writeError(` - error #${i};  ${EOL}`);
          }
          return Promise.resolve(TaskStatus.Failure);
        }
      });
      return taskRunner
        .execute()
        .then(() => fail(EXPECTED_FAIL))
        .catch(err => {
          expect(err.message).toMatchSnapshot();
          expect(terminalProvider.getOutput())
            .toMatch(/List of errors:\S.* - error #1;\S.*lines omitted.* - error #48;\S.* - error #50;\S/);
          checkConsoleOutput(terminalProvider);
        });
    });
  });
});
