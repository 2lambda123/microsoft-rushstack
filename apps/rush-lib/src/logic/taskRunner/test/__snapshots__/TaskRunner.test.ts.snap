// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskRunner Constructor throwsErrorOnInvalidParallelism 1`] = `"Invalid parallelism value of 'tequila', expected a number or 'max'"`;

exports[`TaskRunner Error logging preservedLeadingBlanksButTrimmedTrailingBlanks 1`] = `"Project(s) failed"`;

exports[`TaskRunner Error logging printedAbridgedStdoutAfterErrorWithEmptyStderr 1`] = `"Project(s) failed"`;

exports[`TaskRunner Error logging printedStderrAfterError 1`] = `"Project(s) failed"`;

exports[`TaskRunner Error logging printedStdoutAfterErrorWithEmptyStderr 1`] = `"Project(s) failed"`;

exports[`TaskRunner Warning logging Fail on warning Logs warnings correctly 1`] = `"An error occurred."`;

exports[`TaskRunner Warning logging Fail on warning Logs warnings correctly 2`] = `
Array [
  Object {
    "stream": "O",
    "text": "Executing a maximum of 1 simultaneous processes...[r][n][n]",
  },
  Object {
    "stream": "O",
    "text": "[x][37m[success with warnings (failure)] started[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "Build step 1[r][n][n]",
  },
  Object {
    "stream": "O",
    "text": "Warning: step 1 succeeded with warnings[r][n][n]",
  },
  Object {
    "stream": "E",
    "text": "1 of 1: [success with warnings (failure)] completed with warnings in 0.10 seconds[n]",
  },
  Object {
    "stream": "E",
    "text": "[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m[x][4mSUCCESS WITH WARNINGS (1)[x][24m[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m================================[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m[x][4msuccess with warnings (failure) (0.10 seconds)[x][24m[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m================================[x][39m[n][x][33m[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[n]",
  },
  Object {
    "stream": "E",
    "text": "Project(s) succeeded with warnings[n]",
  },
]
`;

exports[`TaskRunner Warning logging Success on warning Logs warnings correctly 1`] = `
Array [
  Object {
    "stream": "O",
    "text": "Executing a maximum of 1 simultaneous processes...[r][n][n]",
  },
  Object {
    "stream": "O",
    "text": "[x][37m[success with warnings (success)] started[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "Build step 1[r][n][n]",
  },
  Object {
    "stream": "O",
    "text": "Warning: step 1 succeeded with warnings[r][n][n]",
  },
  Object {
    "stream": "E",
    "text": "1 of 1: [success with warnings (success)] completed with warnings in 0.10 seconds[n]",
  },
  Object {
    "stream": "E",
    "text": "[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m[x][4mSUCCESS WITH WARNINGS (1)[x][24m[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m================================[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m[x][4msuccess with warnings (success) (0.10 seconds)[x][24m[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[x][33m================================[x][39m[n][x][33m[x][39m[n]",
  },
  Object {
    "stream": "O",
    "text": "[n]",
  },
]
`;
